### DISCUSSION TITLE
Apple Silicon Wheels

### DISCUSSION THREAD

**Discussion by henryiii** [pypa/cibuildwheel#579] [Show and tell]
I'd like to start this thread to collect user's experiences in adding Apple Silicon wheels, either Universal2 or Arm64, using the new cibuildwheel 1.9.0 and Pip 21.0.1 (at least, that one is needed for Universal2 on AS).

**Comment by henryiii** [comment]
I now ship Universal2 wheels for boost-histogram; I've added them for 0.12.0 by back porting the CI change and manually uploading the new wheel. My CI file is [here](https://github.com/scikit-hep/boost-histogram/blob/develop/.github/workflows/wheels.yml), and the PR adding it was https://github.com/scikit-hep/boost-histogram/pull/495. It worked beautifully the first time; initial performance test look like it's faster on a Mac Mini than on my 8 core i9, maybe by as much as 2x when multithreading (_very_ initial testing, though!). Added it took very little change, and I later also split the lengthy Windows build into auto32/auto64, to keep all the jobs similar in length. Very elegant, I'd say!

**Reply by Czaki** [reply]
> initial performance test look like it's faster on a Mac Mini than on my 8 core i9, maybe by as much as 2x when multithreading

Did you try to compare with wheels compiled against 10.15, not 10.9? Compatybility with old system version could reduce speed. 
I would really see comparision between x86 compiled against current processor and system (maybe also on Ubuntu) and arm64 . 

My base knowledge about compilers and difference of programs speeds on HPC, when using gereral binary and compiled for given machine suggest me that part of these speedup is no need of backward compatybility, and this profit may be lost in few next years (because of a new processors).

**Reply by henryiii** [reply]
No, it was a 10.9+ wheel in both cases, but of course, the AS part of the 10.9 wheel is 11.0+. I'll try a full 11+ on Intel in the near future, excellent suggestion! I'll also try using the universal wheel on Intel, rather than using the normal x86 one. I should get my own AS Mac on Thursday so doing comparisons will be easier then. One benefit of a new architecture is the fact there is no baggage like older systems or even older Python (it's 3.9 only at the moment); it takes many years to drop an old operating system (10.9 is what, 7 years old now?)

**Reply by Czaki** [reply]
Also, it could be interesting to test compilation with `gcc` and usage of `march` flag https://www.alibabacloud.com/blog/gcc-vs--clangllvm-an-in-depth-comparison-of-cc%2B%2B-compilers_595309

Most probably I will have no access to M1 hardware in a short time perspective.

**Comment by YannickJadoul** [comment]
I'm having a bit of a hiccup, adding this to Parselmouth: https://github.com/YannickJadoul/Parselmouth/runs/1845996721

```
Repairing wheel...
  
  + delocate-listdeps /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheela86aupic/built_wheel/praat_parselmouth-0.4.0.dev0-cp39-cp39-macosx_10_9_x86_64.whl && delocate-wheel --require-archs arm64 -w /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheela86aupic/repaired_wheel /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheela86aupic/built_wheel/praat_parselmouth-0.4.0.dev0-cp39-cp39-macosx_10_9_x86_64.whl
                                                              ✓ 3.61s
```

`delocate` seems to be accepting things it shouldn't, here?

**Reply by henryiii** [reply]
It seems to be picking up the existing x86_64 wheel? When I added it, I added it to a separate job (and I didn't build a dedicated arm64 wheel), so might not have seen this.

**Reply by YannickJadoul** [reply]
Hmmm, the wheel building also seems to be producing `x86_64`, so I don't think `delocate` is picking up on the existing wheel. But it is making a mistake in saying "yes, this is an `arm64` wheel" ?

But yes, I now split up things. There might also be a part confusion because I build the largest part of the library (not dependent on Python) just once, in `CIBW_BEFORE_ALL`.

**Reply by YannickJadoul** [reply]
Aha, wait. I see these architecture flags aren't set in `CIBW_BEFORE_ALL`? :-/

**Reply by mayeut** [reply]
If I remember correctly, `--require-archs` only applies to dependencies being grafted.
Nevertheless, IMHO, `cibuildwheel` shall not call `delocate` with those arguments and `delocate` shall reject this.

How did you get `cibuildwheel` to generate this command-line 

For a single `cibuildwheel` run I have:
```
delocate-listdeps /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheele59won_9/built_wheel/pybase64-1.1.4-cp39-cp39-macosx_10_9_x86_64.whl && delocate-wheel --require-archs x86_64 -w /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheele59won_9/repaired_wheel /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheele59won_9/built_wheel/pybase64-1.1.4-cp39-cp39-macosx_10_9_x86_64.whl
```
and
```
delocate-listdeps /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheele59won_9/built_wheel/pybase64-1.1.4-cp39-cp39-macosx_11_0_arm64.whl && delocate-wheel --require-archs arm64 -w /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheele59won_9/repaired_wheel /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibuildwheele59won_9/built_wheel/pybase64-1.1.4-cp39-cp39-macosx_11_0_arm64.whl
```

Could it be that CMake overrides some of `cibuildwheel` environment variable set to build for `arm64` ?

**Reply by YannickJadoul** [reply]
@mayeut, I expect the issue is related to me configuring a subproject in `CIBW_BEFORE_ALL`? As far as I know, I'm not overriding anything else. But maybe CMake has still issues?

Splitting up didn't work, anyway :-/ https://github.com/YannickJadoul/Parselmouth/runs/1846352637 Guess I'll start a debug build to see what's going on.

**Reply by joerick** [reply]
> Aha, wait. I see these architecture flags aren't set in `CIBW_BEFORE_ALL`? :-/

Indeed, they're not.  We actually an issue related to this recently, too: https://github.com/joerick/cibuildwheel/issues/563.

I'm wondering why the wheel ends up with the suffix `x86_64` - what decides the filename? I had assumed that it was overridden by `_PYTHON_HOST_PLATFORM` - apparently not...

**Reply by YannickJadoul** [reply]
> Guess I'll start a debug build to see what's going on.

https://github.com/YannickJadoul/Parselmouth/actions/runs/543933998

Also trying out what happens when I set `ARCHFLAGS`:

https://github.com/YannickJadoul/Parselmouth/actions/runs/543939268

But maybe I should just move this "pre-build" to `CIBW_BEFORE_BUILD` now, yes. It seems that I'm also missing `MACOSX_DEPLOYMENT_TARGET`, yes! I hadn't really considered that before, either :-/

**Reply by YannickJadoul** [reply]
@henryiii, do you know if scikit-build supports Apple Silicon?

**Reply by henryiii** [reply]
The cmake Python package does not come in a universal wheel form. I've been intending to make a PR to do that (since CMake itself does come in a Universal form), but haven't done it yet. See https://github.com/scikit-build/cmake-python-distributions/pull/123#issuecomment-748180984 .

**Reply by YannickJadoul** [reply]
It also passes `-DCMAKE_OSX_ARCHITECTURES=x86_64` if there's no custom `CMAKE_OSX_ARCHITECTURES` option passed in `cmake_args` :-| That's why it still generates `x86_64` wheel names.

**Reply by henryiii** [reply]
Open an issue on scikit-build; I'd offer to help @jcfr with this but with a baby coming, it probably will be a little while till I can help. :(

**Reply by YannickJadoul** [reply]
Thanks! And no worries. I guess I'll just revert to the thing that was +- working before, with your example from pybind11 :-)

I already had this VS 2019 workaround for scikit-build, so things are really not optimal, anyway :-/ https://github.com/scikit-build/scikit-build/pull/526/

**Reply by henryiii** [reply]
For the cmake_exmaple, you'll need to be careful if we enable build, due to [this bug](https://github.com/pypa/build/issues/214) in build (maybe it will be fixed by then). The fact that CMake is not available for ARM/Universal actually shouldn't matter, though we probably want a version of CMake new enough to handle it.

**Reply by YannickJadoul** [reply]
I might have something that works. At least, let's say it produces something with the right wheel file names. See https://github.com/YannickJadoul/Parselmouth/commit/f3b61e00d68995af6d224025a9e29bc84200ed2b and https://github.com/YannickJadoul/Parselmouth/runs/1846906603

Thing I had to change:
- Added `MACOSX_DEPLOYMENT_TARGET` to `CIBW_ENVIRONMENT`. Could be fixed by a solution to #563.
- Added `CMAKE_OSX_ARCHITECTURES` to `CIBW_ENVIRONMENT`. It is kind of unfortunate that CMake doesn't seem to pick up on `ARCHFLAGS` (or maybe it would, but it gets overridden by `CMAKE_OSX_ARCHITECTURES` passed by scikit-learn). We could of course set this environment variable (and I'd be happy with that), but this is also tricky to add something very build-system specific?
    - `ARCHFLAGS`/`CMAKE_OSX_ARCHITECTURES` is also harder/impossible to fix for `CIBW_BEFORE_ALL` in #563, btw, since it's dependent on the Python version.
- Teach scikit-build about the `universal2` <-> `CMAKE_OSX_ARCHITECTURES=x86_64;arm64` link, and teach it to respect `_PYTHON_HOST_PLATFORM` if that's already set (so, yeah, scikit-build is to blame here, @joerick): https://github.com/YannickJadoul/scikit-build/compare/vs2019...YannickJadoul:vs2019-apple-silicon

@henryiii, how do you judge about the scikit-build changes? They seem reasonably ad-hoc and very brittle, so not I'm wondering what else I missed. Also, worth a PR on scikit-build, according to you?

@henryiii, @joerick, @mayeut: any chance one of you would be willing and have time to give these wheels a spin to see if they really are `arm64`/`universal2`? I don't have an Apple Silicon mac at hand (and actually, currently don't even have access to a normal/old mac).

Alternatively, @Czaki, do you know how to inspect these wheels? Can I easily check the `MACOSX_DEPLOYMENT_TARGET` to be 10.9 and the advertised architectures on a Linux machine?

**Reply by Czaki** [reply]
> Alternatively, @Czaki, do you know how to inspect these wheels? Can I easily check the `MACOSX_DEPLOYMENT_TARGET` to be 10.9 and the advertised architectures on a Linux machine?

I'm not sure if I good understand your question. It is possible to read this information. I wrote such code for wheel package. https://github.com/pypa/wheel/blob/master/src/wheel/macosx_libfile.py This is pure python code and should work on any system.

**Reply by YannickJadoul** [reply]
Thanks, @Czaki! Such a thing is exactly what I'm looking for, I think :-) `nm` etc don't work on the format, and I couldn't find a Linux tool to check this :-)

**Reply by YannickJadoul** [reply]
Thank you so much, @Czaki!! I managed to use your code to read the produced files, and everything seems to be correct :-) Ready for a release tomorrow, it seems! :-)

**Reply by henryiii** [reply]
I'm happy to test once I get an AS machine and if I'm not off taking care of a baby. :)

**Reply by YannickJadoul** [reply]
> I'm happy to test once I get an AS machine and if I'm not off taking care of a baby. :)

Ah, I thought you already had one (machine, not the baby, yet ;-) )

**Reply by YannickJadoul** [reply]
Anyway, after checking these headers, I'm quite happy and confident things should work (though anyone giving them a spin is welcome, ofc!). Thanks everyone!

Main question that remains is: what do we do for `cibuildwheel`, cfr. `MACOSX_DEPLOYMENT_TARGET` and `CMAKE_OSX_ARCHITECTURES`? Just a warning in the docs/FAQ? Maybe I'll turn this into a separate issue.
And ofc, what should be upstreamed to scikit-build.

**Reply by henryiii** [reply]
Nope, machine due Thursday, baby due Sunday. I begged testing off of @chrisburr for boost-histogram (who did a fantastic job and ran the benchmarks too), which is the only reason I felt safe shipping it. Though if I knew about @Czaki's tool in wheel, I'd probably have felt okay in shipping it too (but for the first one, I'm really glad I did have proof it ran, and ran fast!).

**Reply by YannickJadoul** [reply]
> Nope, machine due Thursday, baby due Sunday.

That leaves only little time to play around with the new machine ;-)

Yeah, I think I'm good. Apparently my old wheels (advertised as 10.6) were actually 10.13, anyway, and I never got complaints. And I was never hoping for a full test anyway; just some "does it import", but I think that those headers took my worries away. Though I still feel dirty after all the hacks.

**Reply by joerick** [reply]
I just downloaded and installed your universal2 wheel, it looks good. The snag is that I had to install it with `--no-deps`, as numpy doesn't publish a universal2 wheel yet. Naturally it doesn't import either, but fails with `ImportError: ModuleNotFoundError: No module named 'numpy'`. But it looks good to manual inspection:

```shell
$ unzip praat_parselmouth-0.4.0.dev0-cp39-cp39-macosx_10_9_universal2.whl 
$ file parselmouth.cpython-39-darwin.so
parselmouth.cpython-39-darwin.so: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit bundle x86_64] [arm64]
parselmouth.cpython-39-darwin.so (for architecture x86_64):	Mach-O 64-bit bundle x86_64
parselmouth.cpython-39-darwin.so (for architecture arm64):	Mach-O 64-bit bundle arm64
```

**Reply by joerick** [reply]
> Main question that remains is: what do we do for `cibuildwheel`, cfr. `MACOSX_DEPLOYMENT_TARGET` and `CMAKE_OSX_ARCHITECTURES`? Just a warning in the docs/FAQ? Maybe I'll turn this into a separate issue.

It seems like setting `MACOSX_DEPLOYMENT_TARGET` in before_all is a clean win. I wonder if there's any value in us exporting e.g. `CIBUILDWHEEL_BUILD_IDENTIFIER` that would make setting things like `CMAKE_OSX_ARCHITECTURES` a bit cleaner? How do you decide that currently? 

> ARCHFLAGS/CMAKE_OSX_ARCHITECTURES is also harder/impossible to fix for CIBW_BEFORE_ALL in #563, btw, since it's dependent on the Python version.

I was wondering if the natural thing would be to build 'fat' universal2 shared libraries in the BEFORE_ALL step. But maybe that would inflate your binary size too much?

**Reply by YannickJadoul** [reply]
Thanks, @joerick!! :-) Give me even more confidence; it should be good, then, I guess (especially since the NumPy import is also triggered somewhere internally in pybind11, I think, so the module was getting imported, along the way) :-)

**Reply by YannickJadoul** [reply]
> It seems like setting `MACOSX_DEPLOYMENT_TARGET` in before_all is a clean win. I wonder if there's any value in us exporting e.g. `CIBUILDWHEEL_BUILD_IDENTIFIER` that would make setting things like `CMAKE_OSX_ARCHITECTURES` a bit cleaner? How do you decide that currently?

Yes on `MACOSX_DEPLOYMENT_TARGET`.
To set `CMAKE_OSX_ARCHITECTURES`, I do this based on my GHA matrix (where `x86_64`, `arm64`, and `universal2` are split up), such that I know there's only 1 architecture. It's kind of inevitable to rebuild everything, anyway, so I don't mind.

Actually, I'm now realizing I already did this naturally for Windows. There's also only 1 `CIBW_BEFORE_ALL` for both architectures (but they were already split up for time concerns).
So `CIBW_BEFORE_ARCH` would provide another, more principled solution to build non-Python binary dependencies, but... not sure if I'm a big fan of introducing yet another variable, either :-/

> I was wondering if the natural thing would be to build 'fat' universal2 shared libraries in the BEFORE_ALL step. But maybe that would inflate your binary size too much?

Yeah, so you mean to include a fat dependency in `arm64` and `x86_64`? I suppose that would work, but it feels reasonably wasteful, indeed. Most of the code is actually in this external dependency that's independent from Python, and indeed, the universal2 wheel size already doubled (to 19 MB), now.

**Reply by YannickJadoul** [reply]
> > Alternatively, @Czaki, do you know how to inspect these wheels? Can I easily check the `MACOSX_DEPLOYMENT_TARGET` to be 10.9 and the advertised architectures on a Linux machine?
> 
> I'm not sure if I good understand your question. It is possible to read this information. I wrote such code for wheel package. https://github.com/pypa/wheel/blob/master/src/wheel/macosx_libfile.py This is pure python code and should work on any system.

... I now figured out that for the architectures (not `MACOSX_DEPLOYMENT_TARGET`), just a simple `file` works as well :-|

```
$ file parselmouth.cpython-39-darwin.so 
parselmouth.cpython-39-darwin.so: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit x86_64 bundle, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|WEAK_DEFINES|BINDS_TO_WEAK>] [arm64]
```

**Reply by henryiii** [reply]
> The snag is that I had to install it with `--no-deps`, as numpy doesn't publish a universal2 wheel yet.

I would assume many users would use brew (which recently announced official AS support) instead of downloading from python.org, and NumPy should be available there. The downside is you'd have to use `--system-site-packages` to use the system NumPy if you make a venv. I'd love to seem NumPy include Universal2 and/or arm64 wheels!

**Reply by henryiii** [reply]
I got my M1 today. First thing I noticed: python 2.7.16 is preinstalled, and installing the command line tools adds Python 3.8.2. Neither of these actually officially support AS. And while python doesn't come with pip, python3 comes with 19.2.3.

**Reply by YannickJadoul** [reply]
> python 2.7.16 is preinstalled

What the hell...

**Comment by henryiii** [comment]
Interesting, here a project is adding 3.8 (native arm64 cross-compile) wheels outside of cibuildwheel using the Xcode Python 3.8: https://github.com/indygreg/python-zstandard/pull/137

**Reply by Czaki** [reply]
Maybe write them that it works currently?

**Reply by joerick** [reply]
Huh. I wonder if they have a specific use case. Btw, I can confirm that the Python 3.8 that's distributed by Apple inside the Xcode toolchain is universal:

```
$ /Applications/Xcode.app/Contents/Developer/usr/bin/python3 --version
Python 3.8.2
$ /Applications/Xcode.app/Contents/Developer/usr/bin/python3 -c 'import platform; print(platform.machine())'
arm64
```

I wonder if Apple had to make their own modifications to get it to build?

> Maybe write them that it works currently?

We don't do universal builds on Python3.8, yet, because there is no official Python.org universal2 release of Python3.8.

**Reply by henryiii** [reply]
> wonder if they have a specific use case

Probably supporting the default version of CPython on Big Sur? Sadly, we probably can't make this 10.9+ (or maybe even universal2 - 11.0 arm64 would not be unreasonable), and it's quite different from the normal procedure, so likely hard to integrate in. However, if we did add a "use existing Python" mode that has been thrown around before, maybe it would be possible in the future?

> is universal

I'm guessing universal and not arm64 only; I noticed Ruby at least is clearly universal.

**Reply by joerick** [reply]
I hadn't realised but yeah, you're right, `/usr/bin/python3` on Big Sur is a universal2 Python3.8.

```
$ /usr/bin/python3 -c 'import platform; print(platform.machine())'
arm64

$ file /usr/bin/python3
/usr/bin/python3: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64e:Mach-O 64-bit executable arm64e]
/usr/bin/python3 (for architecture x86_64):	Mach-O 64-bit executable x86_64
/usr/bin/python3 (for architecture arm64e):	Mach-O 64-bit executable arm64e

$  /usr/bin/python3 --version                                
Python 3.8.2
```

But yeah, I agree, it's probably not worth us supporting it, with it being such a diversion from our existing workflow.
